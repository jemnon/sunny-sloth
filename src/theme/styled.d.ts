import 'styled-components';
import {
  Breakpoints,
  Colors,
  ColorMode,
  FocusColors,
  Fonts,
  FontSizes,
  Header,
  HoverColors,
  MaxWidths,
  Spacing,
  ZIndex,
} from '../types/theme';

declare module 'styled-components' {
  export interface DefaultTheme {
    breakpoints: Breakpoints;
    colors: Colors;
    focusColors: FocusColors;
    fonts: Fonts;
    fontSizes: FontSizes;
    header: Header;
    hoverColors: HoverColors;
    maxWidths: MaxWidths;
    spacing: Spacing;
    transition: string;
    zIndex: ZIndex;
  }
}
