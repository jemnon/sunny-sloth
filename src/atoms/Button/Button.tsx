import styled, { css } from 'styled-components';
import { Colors } from '../../types/theme';

type ColorScheme = keyof Pick<Colors, 'orange' | 'teal' | 'nearBlack'>;
type Shape = 'pill' | 'rectangle';
type Size = 'small' | 'medium';
type Variant = 'solid' | 'outline';

interface ButtonProps {
  colorScheme?: ColorScheme;
  isDisabled?: boolean;
  isLoading?: boolean;
  variant?: Variant;
  size?: Size;
  shape?: Shape;
  width?: string;
}

export const ButtonReset = css`
  padding: 0;
  border-style: none;
  border-width: 0;
  outline: 0;
  background-color: transparent;
  cursor: pointer;
  text-align: left;
`;

const Button = styled.button<ButtonProps>`
  ${ButtonReset};

  display: inline-flex;
  align-items: center;
  justify-content: center;

  padding: ${({ size = 'medium', theme }): string => {
    if (size === 'small') {
      return `${theme.spacing.sm2} ${theme.spacing.md2}`;
    }
    return `${theme.spacing.sm3} ${theme.spacing.md4}`;
  }};
  min-width: 10rem;
  width: ${({ width = 'auto' }): string => width};

  font-family: ${({ theme }): string => theme.fonts.lato};
  font-size: ${({ size = 'medium', theme }): string =>
    size === 'small' ? '0.85rem' : theme.fontSizes.f1};
  font-weight: bold;
  letter-spacing: normal;
  text-decoration: none;

  border-radius: ${({ shape = 'rounded' }): string =>
    shape === 'rectangle' ? '2px' : '5rem'};
  border-width: ${({ variant = 'solid' }): string =>
    variant === 'outline' ? '1px' : ''};
  border-style: solid;
  border-color: ${({ colorScheme = 'orange', theme }): string =>
    theme.colors[colorScheme]};

  background-color: ${({
    colorScheme = 'orange',
    theme,
    variant = 'solid',
  }): string =>
    variant === 'outline' ? 'transparent' : theme.colors[colorScheme]};
  color: ${({ colorScheme = 'orange', theme, variant = 'solid' }): string =>
    variant === 'outline' ? theme.colors[colorScheme] : theme.colors.white};

  overflow: hidden;
  cursor: ${({ isDisabled, isLoading }): string =>
    isDisabled || isLoading ? 'not-allowed' : 'pointer'};

  transition: box-shadow 0.15s ease;
  transition: background-color 0.47s ease;

  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;

  ${({ isDisabled = false, isLoading = false }): string =>
    isDisabled || isLoading ? 'opacity: 50%;' : ''};

  &:disabled {
    transform: none;
    cursor: not-allowed;
  }
  &:hover {
    ${({ colorScheme = 'orange', theme, variant = 'solid' }): string =>
      variant === 'outline'
        ? `opacity: 0.75`
        : `background-color: ${theme.hoverColors[colorScheme]}`};
  }
  &:active {
    box-shadow: ${({ isDisabled, isLoading, theme }): string =>
      !isDisabled || !isLoading
        ? `${theme.focusColors.blue} 0px 0px 0px 3px`
        : ''};
  }
  & + & {
    margin-left: ${({ theme }): string => theme.spacing.sm4};
  }
`;

export default Button;
