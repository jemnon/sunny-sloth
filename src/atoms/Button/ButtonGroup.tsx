import React, { FC, ReactNode } from 'react';
import styled from 'styled-components';
import {
  generateResponsiveProps,
  CustomInterpolation,
  NonThemeProp,
  SpacingProp,
  ThemeObject,
} from '../../utils/utils';
import { BaseProps } from '../../types/base-props';

type Flow = 'row' | 'column';

export interface ButtonGroupProps extends BaseProps {
  children: ReactNode;
  flow?: Flow;
  sp?: SpacingProp;
  w?: NonThemeProp;
}

const ButtonGroupBase = styled.div<Pick<ButtonGroupProps, 'flow' | 'sp' | 'w'>>`
  display: flex;
  // TODO: figure out resonsive props for flow
  flex-direction: ${({ flow = 'row' }): string => flow};
  ${({ w = '100%' }): CustomInterpolation =>
    generateResponsiveProps<NonThemeProp>({
      cssProperty: 'width',
      prop: w,
    })};
  > * + * {
    ${({ flow = 'row', sp = 'md1', theme: { spacing } }): CustomInterpolation =>
      generateResponsiveProps<SpacingProp>({
        cssProperty: flow === 'row' ? 'margin-left' : 'margin-top',
        prop: sp,
        themeObject: spacing as ThemeObject,
      })}
  }
`;

const ButtonGroup: FC<ButtonGroupProps> = ({
  children,
  dataId,
  flow = 'row',
  id,
  sp = 'md1',
  w = 'auto',
}) => (
  <ButtonGroupBase data-id={dataId} flow={flow} id={id} sp={sp} w={w}>
    {children}
  </ButtonGroupBase>
);

export default ButtonGroup;
