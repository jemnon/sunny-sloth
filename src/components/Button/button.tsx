import styled from 'styled-components';

interface ButtonProps {
  borderColor: string;
  color: string;
  isDisabled: boolean;
  textColor: string;
  type: string;
}

const Button = styled.button<ButtonProps>`
  border: 1px solid
    ${({ theme, borderColor }): string => theme.colors[borderColor]};
  outline: none;
  @media ${({ theme }): string => theme.breakpoints.desktop} {
    padding: 0.75rem 1.25rem;
  }
  padding: 0.5rem;
  font-size: 0.875rem;
  font-weight: bold;
  letter-spacing: 0.1em;
  color: ${({ theme, textColor }): string => theme.colors[textColor]};
  background-color: ${({ theme, color }): string => theme.colors[color]};
  grid-area: button;
  opacity: ${({ isDisabled }): string => (isDisabled ? '0.45' : '1')};
  cursor: ${({ isDisabled }): string =>
    isDisabled ? 'not-allowed' : 'pointer'};
  transition: 0.3s ease opacity;
  &:hover {
    opacity: 0.75;
  }
  &:focus {
    opacity: 0.55;
  }
`;

export default Button;
