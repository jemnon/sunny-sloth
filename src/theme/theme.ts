const theme = {
  breakpoints: {
    sm: '480px',
    md: '960px',
    lg: '1280px',
    xlg: '1440px',
  },
  fonts: {
    latoBold:
      `"Lato-Bold", -apple-system, BlinkMacSystemFont, 'avenir next', avenir, 'Helvetica Neue', ` +
      `helvetica, ubuntu, roboto, noto, 'Segoe UI', Arial, sans-serif`,
    lato:
      `"Lato", -apple-system, BlinkMacSystemFont, 'avenir next', avenir, 'Helvetica Neue', ` +
      `helvetica, ubuntu, roboto, noto, 'Segoe UI', Arial, sans-serif`,
    latoItalic:
      `"Lato-Italic", -apple-system, BlinkMacSystemFont, 'avenir next', avenir, 'Helvetica Neue', ` +
      `helvetica, ubuntu, roboto, noto, 'Segoe UI', Arial, sans-serif`,
    noto:
      `"Noto Serif TC", -apple-system, BlinkMacSystemFont, 'avenir next', avenir, 'Helvetica Neue', ` +
      `helvetica, ubuntu, roboto, noto, 'Segoe UI', Arial, sans-serif`,
    notoBold:
      `"Noto Serif TC-Bold", -apple-system, BlinkMacSystemFont, 'avenir next', avenir, 'Helvetica Neue', ` +
      `helvetica, ubuntu, roboto, noto, 'Segoe UI', Arial, sans-serif`,
  },
  colors: {
    cream: '#f4ede6',
    darkTeal: '#162936',
    teal: '#244c53',
    lightBrown: '#d5a188',
    nearBlack: '#111',
    nearWhite: '#F9F9F9',
    darkOrange: '#BB5B34',
    orange: '#cf7651',
    white: '#ffffff',
    red: '#ff4136',
    green: '#19a974',
  },
  maxWidths: {
    sm: '480px',
    md: '960px',
    lg: '1280px',
  },
  spacing: {
    sm1: '0.25rem' /* 4px */,
    sm2: '0.5rem' /*  8px */,
    sm3: '0.75rem' /* 12px */,
    sm4: '1rem' /* 16px */,
    md1: '1.25rem' /* 20px */,
    md2: '1.5rem' /* 24px */,
    md3: '1.75rem' /* 28px */,
    md4: '2rem' /* 32px */,
    lg1: '2.25rem' /* 36px */,
    lg2: '2.5rem' /* 40px */,
    lg3: '2.75rem' /* 44px */,
    lg4: '3rem' /* 48px */,
  },
  transition: 'all 0.47s ease',
  zIndex: {
    z0: '0',
    z1: '1',
    z2: '2',
    z3: '3',
    z4: '4',
    z5: '5',
    z6: '6',
    z999: '999',
    z9999: '9999',
  },
};

export default theme;
