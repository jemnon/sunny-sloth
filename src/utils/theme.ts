import { DefaultTheme } from 'styled-components';

const theme: DefaultTheme = {
  breakpoints: {
    /* TODO: remove once conversion to styled-breakpoints is done */
    desktop: 'screen and (min-width: 960px)',
    tablet: 'screen and (min-width: 480px) and (max-width: 959px)',
    mobile: 'screen and (max-width: 479px)',
    /* ^^ */
    sm: '480px',
    md: '960px',
    lg: '1440px',
    xlg: '1920px',
  },
  fonts: {
    latoBold:
      `"Lato Bold", -apple-system, BlinkMacSystemFont, 'avenir next', avenir, 'Helvetica Neue', ` +
      `helvetica, ubuntu, roboto, noto, 'Segoe UI', Arial, sans-serif`,
    lato:
      `"Lato", -apple-system, BlinkMacSystemFont, 'avenir next', avenir, 'Helvetica Neue', ` +
      `helvetica, ubuntu, roboto, noto, 'Segoe UI', Arial, sans-serif`,
    latoItalic:
      `"Lato Italic", -apple-system, BlinkMacSystemFont, 'avenir next', avenir, 'Helvetica Neue', ` +
      `helvetica, ubuntu, roboto, noto, 'Segoe UI', Arial, sans-serif`,
    noto:
      `"Noto Serif TC", -apple-system, BlinkMacSystemFont, 'avenir next', avenir, 'Helvetica Neue', ` +
      `helvetica, ubuntu, roboto, noto, 'Segoe UI', Arial, sans-serif`,
  },
  colors: {
    cream: '#f4ede6',
    darkTeal: '#162936',
    teal: '#244c53',
    lightBrown: '#d5a188',
    nearBlack: '#111',
    nearWhite: '#F9F9F9',
    darkOrange: '#BB5B34',
    orange: '#cf7651',
    white: '#ffffff',
    red: '#ff4136',
    green: '#19a974',
  },
  maxWidths: {
    sm: '480px',
    md: '960px',
    lg: '1440px',
  },
  spacing: {
    s1: '0.25rem' /* 4px */,
    s2: '0.5rem' /*  8px */,
    s3: '0.75rem' /* 12px */,
    s4: '1rem' /* 16px */,
    s5: '1.25rem' /* 20px */,
    s6: '1.5rem' /* 24px */,
    s7: '1.75rem' /* 28px */,
    s8: '2rem' /* 32px */,
    s9: '2.25rem' /* 36px */,
    s10: '2.5rem' /* 40px */,
    s11: '2.75rem' /* 44px */,
    s12: '3rem' /* 48px */,
  },
  transition: 'all 0.47s ease',
  zIndex: {
    z0: '0',
    z1: '1',
    z2: '2',
    z3: '3',
    z4: '4',
    z5: '5',
    z6: '6',
    z999: '999',
    z9999: '9999',
  },
};

export default theme;
