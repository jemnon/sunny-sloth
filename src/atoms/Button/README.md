# Button

### Description
Component description

```js
import Button from '@fenderdigital/ui/FILL IN TYPEButton';
```

### Props
* `propName` - property name and description 
