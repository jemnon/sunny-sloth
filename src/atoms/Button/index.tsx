import Button, { ButtonReset } from './Button';
import ButtonGroup from './ButtonGroup';

export { ButtonGroup, ButtonReset };
export default Button;
