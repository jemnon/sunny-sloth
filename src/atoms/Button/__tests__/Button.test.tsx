import React from 'react';
import renderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';
import Button from '../Button';
import theme from '../../../theme';

describe('<Button />', (): void => {
  const component = renderer.create(
    <ThemeProvider theme={theme}>
      <Button />
    </ThemeProvider>
  );
  it('exists', (): void => {
    expect(Button).toBeDefined();
  });
  it('renders defaults correctly', (): void => {
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});