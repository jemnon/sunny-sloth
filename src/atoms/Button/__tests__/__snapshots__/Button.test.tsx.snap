// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> renders defaults correctly 1`] = `
<button
  className="sc-bdfBwQ iETXlT"
/>
`;
